File: Hough-Transform/hough_lines.cpp
#include "hough_common.hpp"

#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <cmath>
#include <filesystem>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>

#ifndef HOUGH_PROJECT_DIR
#error "HOUGH_PROJECT_DIR must be defined by CMake"
#endif

namespace {

int parseInteger(const std::string& text) {
    std::size_t consumed = 0;
    const int value = std::stoi(text, &consumed);
    if (consumed != text.size()) {
        throw std::invalid_argument("invalid integer: " + text);
    }
    return value;
}

struct Options {
    std::filesystem::path input =
        std::filesystem::path(HOUGH_PROJECT_DIR) / "lanes.jpg";
    std::optional<std::filesystem::path> outputDirectory;
    int cannyLow = 50;
    int cannyHigh = 150;
    int houghThreshold = 50;
    int minimumLineLength = 40;
    int maximumLineGap = 25;
    bool noDisplay = false;
    bool validate = false;
    bool help = false;
};

Options parseArguments(const int argc, char** argv) {
    Options options;
    bool inputWasSet = false;
    for (int index = 1; index < argc; ++index) {
        const std::string argument(argv[index]);
        if (argument == "--output-dir") {
            if (++index >= argc) {
                throw std::invalid_argument("--output-dir requires a path");
            }
            options.outputDirectory = std::filesystem::absolute(argv[index]);
        } else if (argument == "--canny-low") {
            if (++index >= argc) {
                throw std::invalid_argument("--canny-low requires an integer");
            }
            options.cannyLow = parseInteger(argv[index]);
        } else if (argument == "--canny-high") {
            if (++index >= argc) {
                throw std::invalid_argument("--canny-high requires an integer");
            }
            options.cannyHigh = parseInteger(argv[index]);
        } else if (argument == "--hough-threshold") {
            if (++index >= argc) {
                throw std::invalid_argument("--hough-threshold requires an integer");
            }
            options.houghThreshold = parseInteger(argv[index]);
        } else if (argument == "--min-line-length") {
            if (++index >= argc) {
                throw std::invalid_argument("--min-line-length requires an integer");
            }
            options.minimumLineLength = parseInteger(argv[index]);
        } else if (argument == "--max-line-gap") {
            if (++index >= argc) {
                throw std::invalid_argument("--max-line-gap requires an integer");
            }
            options.maximumLineGap = parseInteger(argv[index]);
        } else if (argument == "--no-display") {
            options.noDisplay = true;
        } else if (argument == "--validate") {
            options.validate = true;
        } else if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else if (!argument.empty() && argument[0] == '-') {
            throw std::invalid_argument("unknown argument: " + argument);
        } else if (!inputWasSet) {
            options.input = std::filesystem::absolute(argument);
            inputWasSet = true;
        } else {
            throw std::invalid_argument("only one input image may be provided");
        }
    }
    return options;
}

bool hasHorizontalLine(
    const std::vector<cv::Vec4i>& lines,
    const int minimumLength = 150) {
    for (const cv::Vec4i& line : lines) {
        if (std::abs(line[3] - line[1]) <= 3 &&
            std::abs(line[2] - line[0]) >= minimumLength) {
            return true;
        }
    }
    return false;
}

bool hasDiagonalLine(
    const std::vector<cv::Vec4i>& lines,
    const double minimumLength = 130.0) {
    for (const cv::Vec4i& line : lines) {
        const int deltaX = line[2] - line[0];
        const int deltaY = line[3] - line[1];
        if (std::hypot(deltaX, deltaY) >= minimumLength &&
            std::abs(deltaX) > 20 &&
            std::abs(deltaY) > 20) {
            return true;
        }
    }
    return false;
}

std::pair<std::filesystem::path, std::filesystem::path> outputPaths(
    const Options& options) {
    if (!options.outputDirectory.has_value()) {
        throw std::logic_error("output directory is unavailable");
    }
    const std::string stem = options.input.stem().string();
    return {
        *options.outputDirectory / (stem + "-edges.png"),
        *options.outputDirectory / (stem + "-lines.png"),
    };
}

void writeOutputs(
    const Options& options,
    const hough::LineDetection& detection,
    const cv::Mat& annotated) {
    const auto [edgePath, resultPath] = outputPaths(options);
    hough::writeImage(edgePath, detection.edges);
    hough::writeImage(resultPath, annotated);
    std::cout << "Wrote: " << edgePath << '\n';
    std::cout << "Wrote: " << resultPath << '\n';
}

void runValidation(const Options& options) {
    cv::Mat synthetic = cv::Mat::zeros(256, 256, CV_8UC3);
    cv::line(
        synthetic,
        cv::Point(20, 45),
        cv::Point(235, 45),
        cv::Scalar(255, 255, 255),
        3,
        cv::LINE_AA);
    cv::line(
        synthetic,
        cv::Point(25, 225),
        cv::Point(225, 75),
        cv::Scalar(255, 255, 255),
        3,
        cv::LINE_AA);
    const hough::LineDetection syntheticDetection =
        hough::detectLines(synthetic, 50, 150, 45, 100, 12);
    if (!hasHorizontalLine(syntheticDetection.lines) ||
        !hasDiagonalLine(syntheticDetection.lines)) {
        throw std::runtime_error("synthetic semantic lines were not recovered");
    }

    const cv::Mat image = hough::readBgr(options.input);
    const hough::LineDetection detection = hough::detectLines(image);
    if (detection.lines.empty()) {
        throw std::runtime_error("no lines were detected in lanes.jpg");
    }
    const cv::Mat annotated = hough::drawLines(image, detection.lines);
    if (options.outputDirectory.has_value()) {
        writeOutputs(options, detection, annotated);
    }

    std::cout << "VALIDATION PASSED: lines=" << detection.lines.size()
              << ", edge_pixels=" << cv::countNonZero(detection.edges)
              << ", synthetic_lines=" << syntheticDetection.lines.size() << '\n';
}

void printUsage(const char* program) {
    std::cout
        << "Usage: " << program
        << " [IMAGE] [--output-dir DIR] [--canny-low N] [--canny-high N]"
        << " [--hough-threshold N] [--min-line-length N] [--max-line-gap N]"
        << " [--no-display] [--validate]\n";
}

}  // namespace

int main(const int argc, char** argv) {
    try {
        const Options options = parseArguments(argc, argv);
        if (options.help) {
            printUsage(argv[0]);
            return 0;
        }
        if (options.validate) {
            runValidation(options);
            return 0;
        }

        const cv::Mat image = hough::readBgr(options.input);
        const hough::LineDetection detection = hough::detectLines(
            image,
            options.cannyLow,
            options.cannyHigh,
            options.houghThreshold,
            options.minimumLineLength,
            options.maximumLineGap);
        const cv::Mat annotated = hough::drawLines(image, detection.lines);
        std::cout << "Input: " << options.input << '\n';
        std::cout << "Line segments: " << detection.lines.size() << '\n';
        std::cout << "Edge pixels: " << cv::countNonZero(detection.edges) << '\n';

        if (options.outputDirectory.has_value()) {
            writeOutputs(options, detection, annotated);
        }
        if (!options.noDisplay) {
            cv::imshow("Edges", detection.edges);
            cv::imshow("Probabilistic Hough lines", annotated);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        printUsage(argv[0]);
        return 2;
    }
}

File: Hough-Transform/hough_lines.py
#!/usr/bin/env python3
"""Detect line segments with Canny edges and the probabilistic Hough transform."""

from __future__ import annotations

import argparse
from collections.abc import Sequence
from pathlib import Path

import cv2
import numpy as np

from hough_utils import (
    detect_lines,
    draw_lines,
    line_output_paths,
    read_bgr,
    resolve_input,
    write_image,
)


def _has_horizontal_line(lines: np.ndarray, minimum_length: int = 150) -> bool:
    """Return True when a long, nearly horizontal semantic line is present."""

    for x1, y1, x2, y2 in lines:
        if abs(int(y2) - int(y1)) <= 3 and abs(int(x2) - int(x1)) >= minimum_length:
            return True
    return False


def _has_diagonal_line(lines: np.ndarray, minimum_length: float = 130.0) -> bool:
    """Return True when a long diagonal semantic line is present."""

    for x1, y1, x2, y2 in lines:
        delta_x = int(x2) - int(x1)
        delta_y = int(y2) - int(y1)
        length = float(np.hypot(delta_x, delta_y))
        if length >= minimum_length and abs(delta_x) > 20 and abs(delta_y) > 20:
            return True
    return False


def run_validation(output_dir: str | Path | None = None) -> dict[str, int]:
    """Validate semantic lines on a synthetic image plus the bundled road image."""

    synthetic = np.zeros((256, 256, 3), dtype=np.uint8)
    cv2.line(synthetic, (20, 45), (235, 45), (255, 255, 255), 3, cv2.LINE_AA)
    cv2.line(synthetic, (25, 225), (225, 75), (255, 255, 255), 3, cv2.LINE_AA)
    _, synthetic_lines = detect_lines(
        synthetic,
        hough_threshold=45,
        min_line_length=100,
        max_line_gap=12,
    )
    if not _has_horizontal_line(synthetic_lines):
        raise AssertionError("synthetic horizontal line was not recovered")
    if not _has_diagonal_line(synthetic_lines):
        raise AssertionError("synthetic diagonal line was not recovered")

    input_path = resolve_input(None, "lanes.jpg")
    image = read_bgr(input_path)
    edges, lines = detect_lines(image)
    if lines.size == 0:
        raise AssertionError("no lines were detected in lanes.jpg")
    annotated = draw_lines(image, lines)

    if output_dir is not None:
        edge_path, result_path = line_output_paths(output_dir, input_path.stem)
        write_image(edge_path, edges)
        write_image(result_path, annotated)

    result = {
        "line_count": int(lines.shape[0]),
        "edge_pixels": int(cv2.countNonZero(edges)),
        "synthetic_line_count": int(synthetic_lines.shape[0]),
    }
    print(
        "VALIDATION PASSED: "
        f"lines={result['line_count']}, "
        f"edge_pixels={result['edge_pixels']}, "
        f"synthetic_lines={result['synthetic_line_count']}"
    )
    return result


def _build_parser() -> argparse.ArgumentParser:
    """Create the line-detector CLI parser."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", nargs="?", type=Path, help="input image; defaults to lanes.jpg")
    parser.add_argument("--output-dir", type=Path, help="directory for edges and line result")
    parser.add_argument("--canny-low", type=int, default=50)
    parser.add_argument("--canny-high", type=int, default=150)
    parser.add_argument("--hough-threshold", type=int, default=50)
    parser.add_argument("--min-line-length", type=int, default=40)
    parser.add_argument("--max-line-gap", type=int, default=25)
    parser.add_argument("--no-display", action="store_true", help="do not open GUI windows")
    parser.add_argument("--validate", action="store_true", help="run deterministic checks")
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    """Detect lines, report metrics, and optionally save/display outputs."""

    arguments = _build_parser().parse_args(argv)
    if arguments.validate:
        run_validation(arguments.output_dir)
        return 0

    input_path = resolve_input(arguments.input, "lanes.jpg")
    image = read_bgr(input_path)
    edges, lines = detect_lines(
        image,
        canny_low=arguments.canny_low,
        canny_high=arguments.canny_high,
        hough_threshold=arguments.hough_threshold,
        min_line_length=arguments.min_line_length,
        max_line_gap=arguments.max_line_gap,
    )
    annotated = draw_lines(image, lines)
    print(f"Input: {input_path}")
    print(f"Line segments: {lines.shape[0]}")
    print(f"Edge pixels: {cv2.countNonZero(edges)}")

    if arguments.output_dir is not None:
        edge_path, result_path = line_output_paths(
            arguments.output_dir, input_path.stem
        )
        write_image(edge_path, edges)
        write_image(result_path, annotated)
        print(f"Wrote: {edge_path}")
        print(f"Wrote: {result_path}")

    if not arguments.no_display:
        cv2.imshow("Edges", edges)
        cv2.imshow("Probabilistic Hough lines", annotated)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
